File: app/src/main/java/io/github/skywalkerdarren/simpleaccounting/view_model/DiscoveryViewModel.java
package io.github.skywalkerdarren.simpleaccounting.view_model;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import org.joda.time.DateTime;

import java.util.List;

import io.github.skywalkerdarren.simpleaccounting.model.AppRepository;
import io.github.skywalkerdarren.simpleaccounting.model.datasource.CurrencyDataSource;
import io.github.skywalkerdarren.simpleaccounting.model.entity.Currency;
import io.github.skywalkerdarren.simpleaccounting.model.entity.CurrencyInfo;

public class DiscoveryViewModel extends ViewModel {
    private final AppRepository mRepository;
    private final MutableLiveData<String> cumulativeDays = new MutableLiveData<>();
    private final MutableLiveData<String> monthOfAccountingCounts = new MutableLiveData<>();
    private final MutableLiveData<String> sumOfAccountingCounts = new MutableLiveData<>();
    private final MutableLiveData<String> monthReport = new MutableLiveData<>();
    private final MutableLiveData<String> yearReport = new MutableLiveData<>();
    private final MutableLiveData<String> weekReport = new MutableLiveData<>();
    private final MutableLiveData<String> currentCurrency = new MutableLiveData<>();
    private final MutableLiveData<String> currentCurrencyPic = new MutableLiveData<>();
    private final MutableLiveData<List<Currency>> favoriteCurrencies = new MutableLiveData<>();

    public DiscoveryViewModel(AppRepository repository) {
        String defCurrency = "CNY";
        mRepository = repository;
        mRepository.getBillsCount(count -> sumOfAccountingCounts.setValue(count + "单"));
        DateTime dateTime = new DateTime();
        mRepository.getBillsCount(dateTime.getYear(), dateTime.getMonthOfYear(),
                count -> monthOfAccountingCounts.setValue(count + "单"));
        monthReport.setValue(dateTime.getMonthOfYear() + "月报表");
        yearReport.setValue(dateTime.getYear() + "年报表");
        weekReport.setValue("第" + dateTime.getWeekOfWeekyear() + "周报表");
        mRepository.getCurrencyInfo(defCurrency, new CurrencyDataSource.LoadCurrencyInfoCallback() {
            @Override
            public void onCurrencyInfoLoaded(CurrencyInfo info) {
                currentCurrency.setValue(info.getFullNameCN());
                currentCurrencyPic.setValue(info.getFlagLocation());
            }

            @Override
            public void onDataUnavailable() {

            }
        });
        mRepository.getFavouriteCurrenciesExchangeRate(defCurrency, new CurrencyDataSource.LoadExchangeRatesCallback() {
            @Override
            public void onExchangeRatesLoaded(List<Currency> currencies) {
                favoriteCurrencies.setValue(currencies);
            }

            @Override
            public void onDataUnavailable() {

            }
        });
    }

    public MutableLiveData<String> getCurrentCurrency() {
        return currentCurrency;
    }

    public MutableLiveData<String> getCurrentCurrencyPic() {
        return currentCurrencyPic;
    }

    public MutableLiveData<String> getCumulativeDays() {
        return cumulativeDays;
    }

    public void setCumulativeDays(String days) {
        cumulativeDays.setValue(days);
    }

    public MutableLiveData<String> getMonthOfAccountingCounts() {
        return monthOfAccountingCounts;
    }

    public MutableLiveData<String> getSumOfAccountingCounts() {
        return sumOfAccountingCounts;
    }

    public MutableLiveData<String> getMonthReport() {
        return monthReport;
    }

    public MutableLiveData<String> getYearReport() {
        return yearReport;
    }

    public MutableLiveData<String> getWeekReport() {
        return weekReport;
    }

    public LiveData<List<Currency>> getFavoriteCurrencies() {
        return favoriteCurrencies;
    }
}

File: app/src/main/java/io/github/skywalkerdarren/simpleaccounting/view_model/DiscoveryViewModel.kt
package io.github.skywalkerdarren.simpleaccounting.view_model

import android.content.Context
import android.util.Log
import android.widget.Toast
import androidx.lifecycle.*
import io.github.skywalkerdarren.simpleaccounting.model.AppRepository
import io.github.skywalkerdarren.simpleaccounting.model.datasource.BillDataSource.LoadBillCountCallBack
import io.github.skywalkerdarren.simpleaccounting.model.entity.CurrenciesInfo
import io.github.skywalkerdarren.simpleaccounting.model.entity.Currency
import io.github.skywalkerdarren.simpleaccounting.model.entity.CurrencyAndInfo
import io.github.skywalkerdarren.simpleaccounting.model.entity.CurrencyInfo
import io.github.skywalkerdarren.simpleaccounting.model.repository.CurrencyRepo
import io.github.skywalkerdarren.simpleaccounting.util.data.PreferenceUtil
import io.github.skywalkerdarren.simpleaccounting.util.network.RequestService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.joda.time.DateTime
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class DiscoveryViewModel(private val mRepository: AppRepository, private val currencyRepo: CurrencyRepo) : ViewModel() {
    val defCurrency: MutableLiveData<String> = MutableLiveData("CNY")
    val cumulativeDays = MutableLiveData<String>()
    val monthOfAccountingCounts = MutableLiveData<String>()
    val sumOfAccountingCounts = MutableLiveData<String>()
    val monthReport = MutableLiveData<String>()
    val yearReport = MutableLiveData<String>()
    val weekReport = MutableLiveData<String>()
    val currentCurrency: LiveData<CurrencyInfo> = Transformations.switchMap(defCurrency) {
        currencyRepo.getCurrencyInfo(it)
    }
    val favoriteCurrencies: LiveData<List<CurrencyAndInfo>> = Transformations.switchMap(defCurrency) {
        currencyRepo.getFavouriteCurrenciesExchangeRate(it)
    }

    val currencyDate = MutableLiveData<String>()

    fun setCurrencyDate(date: String) {
        currencyDate.value = date
    }

    fun setCumulativeDays(days: String) {
        cumulativeDays.value = days
    }

    fun refreshCurrency(context: Context) {
        viewModelScope.launch(Dispatchers.IO) {
            // 上次更新时间戳
            val timestamp = PreferenceUtil.getString(context, PreferenceUtil.LAST_UPDATE_TIMESTAMP, "0")
            val before = DateTime(timestamp.toLong() * 1000)
            val after = DateTime.now()
            if (!after.minusDays(1).isAfter(before)) {
                fail(context, "更新时间间隔至少一天")
                return@launch
            }

            val service = RequestService().requestCurrenciesInfo(context)
            service.enqueue(object : Callback<CurrenciesInfo> {
                override fun onFailure(call: Call<CurrenciesInfo>, t: Throwable) {
                    fail(context, "网络错误")
                    Log.e("wtf", t.message, t)
                }

                override fun onResponse(call: Call<CurrenciesInfo>, response: Response<CurrenciesInfo>) {
                    if (response.isSuccessful) {
                        val currenciesInfo = response.body()
                        if ("false" == currenciesInfo?.success) {
                            fail(context, "网站错误，大概是token错了")
                            Log.e("wtf", currenciesInfo.error.toString())
                            return
                        }
                        val time = currenciesInfo?.timestamp
                        PreferenceUtil.setString(context, PreferenceUtil.LAST_UPDATE_TIMESTAMP, time)
                        val currencies: List<Currency>? = currenciesInfo?.quotes
                        if (currencies == null || time == null) {
                            fail(context, "没有数据，蜜汁问题")
                            Log.e("wtf", "no currency")
                            return
                        }
                        viewModelScope.launch(Dispatchers.IO) end@{
                            currencyRepo.updateCurrencies(currencies)
                            viewModelScope.launch {
                                currencyDate.value = DateTime(time.toLong() * 1000)
                                        .toString("yyyy-MM-dd")
                            }
                        }
                    }
                }

            })
        }

    }

    private fun fail(context: Context, msg: String) {
        viewModelScope.launch(Dispatchers.Main) {
            Toast.makeText(context, msg, Toast.LENGTH_SHORT).show()
        }
    }

    fun start() {
        mRepository.getBillsCount(object : LoadBillCountCallBack {
            override fun onBillCountLoaded(count: Int) {
                sumOfAccountingCounts.value = count.toString() + "单"
            }
        })
        val dateTime = DateTime()
        mRepository.getBillsCount(dateTime.year, dateTime.monthOfYear,
                object : LoadBillCountCallBack {
                    override fun onBillCountLoaded(count: Int) {
                        monthOfAccountingCounts.value = count.toString() + "单"
                    }
                })
        monthReport.value = dateTime.monthOfYear.toString() + "月报表"
        yearReport.value = dateTime.year.toString() + "年报表"
        weekReport.value = "第" + dateTime.weekOfWeekyear + "周报表"
    }

    fun changeCurrency(currencyA: Currency?, currencyB: Currency?) {
        currencyA ?: return
        currencyB ?: return
        viewModelScope.launch(Dispatchers.IO) {
            currencyRepo.changeCurrencyPosition(currencyA, currencyB)
        }
    }
}